### ISSUE TITLE
Update/Add Coursier

### ISSUE THREAD

**Issue by laughedelic** [actions/virtual-environments#3121]
**Tool information**
- Tool name: Coursier (`cs`)
- Tool license: Apache v2
- Add or update? add
- Desired version: [latest stable](https://github.com/coursier/coursier/releases/latest)
- Approximate size: native installer/launcher size is ~55MB (not sure about the full installation size)
- If this is an add request:
  - Brief description of tool: 
    
    Quoting the [official page](https://get-coursier.io/docs/overview):
    > coursier is a dependency resolver / fetcher à la Maven / Ivy, entirely rewritten from scratch in 
    > Scala. It aims at being fast and easy to embed in other contexts. 
    >
    > It handles many features of the Maven model, and is able to fetch metadata and artifacts from both Maven and Ivy repositories. It handles parallel downloads out-of-the-box without resorting to global locks.
 
    As a [CLI tool](https://get-coursier.io/docs/cli-overview) it allows
    > - [easily list the transitive dependencies of applications or libraries,](https://get-coursier.io/docs/cli-resolve)
    > - [download and list their artifacts,](https://get-coursier.io/docs/cli-fetch)
    > - [run applications published via Maven / Ivy repositories,](https://get-coursier.io/docs/cli-launch)
    > - [and many other things.](https://get-coursier.io/docs/cli-overview)
   
    I think its functionality can be compared to that of pip/pipx for Python, npm/npx for Node.js, or Homebrew (in the context of the Scala ecosystem).

    It's especially useful on CI for setting up the project environment: install a particular version of the JVM, install other tools (e.g. [mill](https://com-lihaoyi.github.io/mill), [scalafmt](https://scalameta.org/scalafmt/), [scalafix](https://scalacenter.github.io/scalafix/), [ammonite](http://ammonite.io/)), fetch artifacts, run JVM applications, bootstrap executables, etc. 

  - URL for tool's homepage: https://get-coursier.io
  - Provide a basic test case to validate the tool's functionality: 
      `cs --version` prints out the version

**Area for Triage**:

Scala

**Question, Bug, or Feature?**:

Feature

**Virtual environments affected**

- [x] Ubuntu 16.04
- [x] Ubuntu 18.04
- [x] Ubuntu 20.04
- [x] macOS 10.15
- [x] macOS 11.0
- [x] Windows Server 2016 R2
- [x] Windows Server 2019

**Can this tool be installed during the build?**

Yes. [Official installation instructions](https://get-coursier.io/docs/cli-installation).

**Tool installation time in runtime**

<10s?

**Are you willing to submit a PR?**

Yes 🙌 

----

Related GitHub actions:
- https://github.com/coursier/cache-action
- https://github.com/laughedelic/coursier-setup

/cc @alexarchambault (author of Coursier)

**Comment by AlenaSviridenko** [comment]
Hi @laughedelic,
official GitHub recommendation is using actions to setup tools whenever possible https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#using-preinstalled-software, as it is the most safe and flexible way of using them, so we'd rather not to preinstall this tool to the image.

I am closing this issue, please, let us know if you have any concerns. Thanks.

**Comment by laughedelic** [comment]
Hi @AlenaSviridenko, I would argue that having a preinstalled latest version of Coursier is more important for Scala projects than having sbt (which is already included) because Coursier can be used to install (any version of) sbt or any other Scala build tool (e.g. bloop, mill, seed).

Could you explain the criteria for preinstalling sbt if there are actions (e.g. [setup-scala](https://github.com/marketplace/actions/setup-scala)) that can do it in runtime? 

Also, as I pointed out in the description, Coursier CLI provides functionality comparable to that of pip/pipx, npm/npx and Homebrew which are all preinstalled. Why can't we have such tooling preinstalled for the Scala ecosystem?

**Comment by ahjohannessen** [comment]
Please reconsider this

**Comment by ScalaWilliam** [comment]
As CS is quite lightweight there should not be good reason not to include it - it in fact allows extra possibilities over Maven and SBT for any JVM application, thus enhancing the ecosystem.

**Comment by KadekM** [comment]
Please include the coursier in same manner pip/npm are included

**Comment by AlenaSviridenko** [comment]
Hey,
I see your point here. Indeed, previously we installed tools by making decisions based on tool installation time and size only (like for sbt), but after several months with such approach we started facing with a lot of issues when the tool got upgraded and customers got broken with this update, or when tools are conflicting with each other. 
So we came to the point that whenever a tool has an action and its installation time is pretty small, we insist on the installation with action in runtime, so its prebaking won't affect anyone who is not interested in this tool. Also, actions usually provides wider setup functionality and their maintainers have more expertise in this tool than we have.
